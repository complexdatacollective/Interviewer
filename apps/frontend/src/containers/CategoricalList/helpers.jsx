const getIsPortrait = (width, height) => width / height < 1;

export const getExpandedSize = (bounds) => (
  bounds.height > (bounds.width * 0.67)
    ? Math.floor(bounds.width * 0.67)
    : bounds.height
);

export const getItemSize = (bounds, itemCount, expanded = false) => {
  const expandedSize = getExpandedSize(bounds);

  const width = expanded ? bounds.width - expandedSize : bounds.width;

  const isPortrait = getIsPortrait(width, bounds.height);

  const unexpandedItemCount = expanded ? itemCount - 1 : itemCount;

  const shortCount = [4, 5, 6, 7, 8].includes(unexpandedItemCount) ? 2 : 1;
  const longCount = shortCount > 1 ? Math.ceil(unexpandedItemCount / 2) : unexpandedItemCount;

  const longSide = isPortrait ? bounds.height : width;
  const shortSide = isPortrait ? width : bounds.height;

  const x = Math.floor(longSide / longCount);
  const y = Math.floor(shortSide / shortCount);
  const z = x < y ? x : y;

  return z;
};
