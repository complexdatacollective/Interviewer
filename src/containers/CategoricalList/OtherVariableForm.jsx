import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';
import { Button, Node } from '@codaco/ui';
import { MarkdownLabel } from '@codaco/ui/lib/components/Fields';
import Field from '../Field';

const stopClickPropagation = (e) => e.stopPropagation();

const OtherVariableForm = ({
  otherVariablePrompt,
  handleSubmit,
  color,
  label,
  onCancel,
}) => (
  <div className="other-variable-form" onClick={stopClickPropagation}>
    <form onSubmit={handleSubmit}>
      <div className="other-variable-form__content">
        <div className="other-variable-form__content-left">
          <Node label={label} color={color} />
        </div>
        <div className="other-variable-form__content-right">
          <h4>
            <MarkdownLabel inline label={otherVariablePrompt} />
          </h4>
          <Field
            label=""
            placeholder="Enter your response here..."
            component="Text"
            name="otherVariable"
            validation={{ required: true }}
          />
        </div>
      </div>
      <div className="other-variable-form__footer">
        <Button type="button" color="white" onClick={onCancel}>Cancel</Button>
        <Button type="submit" aria-label="Submit">Continue</Button>
      </div>
    </form>
  </div>
);

OtherVariableForm.propTypes = {
  otherVariablePrompt: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  color: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};

export default reduxForm({ form: 'otherVariableForm' })(OtherVariableForm);
