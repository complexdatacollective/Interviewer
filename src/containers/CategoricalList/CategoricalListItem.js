import React, { useState } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { entityPrimaryKeyProperty } from '@codaco/shared-consts';
import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
import { getEntityAttributes } from '../../ducks/modules/network';
import { CategoricalItem } from '../../components';
import Overlay from '../Overlay';
import OtherVariableForm from './OtherVariableForm';
import { get } from '../../utils/lodash-replacements';

const formatBinDetails = (nodes, getNodeLabel) => {
  if (nodes.length === 0) {
    return '';
  }

  // todo: the following should be updated to reflect the sort order of the bins
  const name = getNodeLabel(nodes[0]);

  return `${name}${nodes.length > 1 ? ` and ${nodes.length - 1} other${nodes.length > 2 ? 's' : ''}` : ''}`;
};

const otherVariableWindowInitialState = {
  show: false,
  node: null,
};

const CategoricalListItem = (props) => {
  const {
    id,
    size,
    isExpanded,
    accentColor,
    activePromptVariable,
    promptOtherVariable,
    bin,
    index,
    sortOrder,
    getNodeLabel,
    nodeColor,
    onExpandBin,
    updateNode,
    stage,
  } = props;

  const isOtherVariable = !!bin.otherVariable;
  const [otherVariableWindow, setOtherVariableWindow] = useState(otherVariableWindowInitialState);
  const binDetails = formatBinDetails(bin.nodes, getNodeLabel);

  const openOtherVariableWindow = (node) => {
    const otherVariable = get(getEntityAttributes(node), bin.otherVariable);

    setOtherVariableWindow({
      show: true,
      node,
      label: getNodeLabel(node),
      color: nodeColor,
      initialValues: {
        otherVariable,
      },
    });
  };

  const closeOtherVariableWindow = () => setOtherVariableWindow(otherVariableWindowInitialState);

  const setNodeCategory = (node, category) => {
    const variable = bin.otherVariable || activePromptVariable;

    const resetVariable = bin.otherVariable ? activePromptVariable : promptOtherVariable;

    // categorical requires an array, otherVariable is a string
    const value = bin.otherVariable ? category : [category];

    if (getEntityAttributes(node)[variable] === value) {
      return;
    }

    updateNode(
      node[entityPrimaryKeyProperty],
      {},
      {
        [variable]: value,
        // reset is used to clear the variable when a node is moved to a different bin
        ...(!!resetVariable && { [resetVariable]: null }),
      },
      'drop',
    );
  };

  const handleDrop = ({ meta: node }) => {
    const binValue = bin.value;

    if (isOtherVariable) {
      openOtherVariableWindow(node);
      return;
    }

    setNodeCategory(node, binValue);
  };

  const handleClickItem = (node) => {
    if (!isOtherVariable) { return; }
    openOtherVariableWindow(node);
  };

  const handleSubmitOtherVariableForm = ({ otherVariable: value }) => {
    const { node } = otherVariableWindow;

    setNodeCategory(node, value);
    closeOtherVariableWindow();
  };

  const handleExpandBin = (e) => {
    if (e) { e.stopPropagation(); }
    onExpandBin(index);
  };

  return (
    <div
      className="categorical-list__item"
      style={{ width: `${size}px`, height: `${size}px` }}
      key={index}
      onClick={handleExpandBin}
    >
      <CategoricalItem
        id={id}
        key={index}
        label={bin.label}
        accentColor={accentColor}
        onDrop={handleDrop}
        onClick={handleExpandBin}
        onClickItem={handleClickItem}
        details={binDetails}
        isExpanded={isExpanded}
        nodes={bin.nodes}
        sortOrder={sortOrder}
        stage={stage}
      />
      {isOtherVariable
        && (
          <Overlay
            show={otherVariableWindow.show}
            onClose={closeOtherVariableWindow}
            onBlur={closeOtherVariableWindow}
          >
            {otherVariableWindow.show
              && (
                <OtherVariableForm
                  label={otherVariableWindow.label}
                  color={otherVariableWindow.color}
                  otherVariablePrompt={bin.otherVariablePrompt}
                  onSubmit={handleSubmitOtherVariableForm}
                  onCancel={closeOtherVariableWindow}
                  initialValues={otherVariableWindow.initialValues}
                />
              )}
          </Overlay>
        )}
    </div>
  );
};

CategoricalListItem.propTypes = {
  id: PropTypes.string.isRequired,
  isExpanded: PropTypes.bool.isRequired,
  activePromptVariable: PropTypes.string.isRequired,
  promptOtherVariable: PropTypes.string,
  bin: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  sortOrder: PropTypes.array,
  getNodeLabel: PropTypes.func.isRequired,
  nodeColor: PropTypes.string,
  onExpandBin: PropTypes.func.isRequired,
  updateNode: PropTypes.func.isRequired,
  accentColor: PropTypes.string,
  size: PropTypes.number,
};

CategoricalListItem.defaultProps = {
  promptOtherVariable: null,
  accentColor: null,
  nodeColor: null,
  size: 0,
  sortOrder: [],
};

export { CategoricalListItem };

const mapDispatchToProps = {
  updateNode: sessionsActions.updateNode,
};

export default compose(
  connect(null, mapDispatchToProps),
)(CategoricalListItem);
