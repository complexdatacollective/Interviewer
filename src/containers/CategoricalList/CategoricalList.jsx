import React, { Component } from 'react';
import { compose } from 'redux';
import { throttle } from 'lodash';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Flipper } from 'react-flip-toolkit';
import cx from 'classnames';
import color from 'color';
import { getCSSVariableAsString } from '@codaco/ui/lib/utils/CSSVariables';
import { entityAttributesProperty } from '@codaco/shared-consts';
import {
  makeNetworkNodesForType, makeGetVariableOptions, makeGetPromptVariable, getPromptOtherVariable,
} from '../../selectors/interface';
import { makeGetNodeLabel, makeGetNodeColor } from '../../selectors/network';
import { MonitorDragSource } from '../../behaviours/DragAndDrop';
import getAbsoluteBoundingRect from '../../utils/getAbsoluteBoundingRect';
import CategoricalListItem from './CategoricalListItem';
import { getItemSize, getExpandedSize } from './helpers';

const isSpecialValue = (value) => {
  if (value === null) {
    return true;
  }
  if (typeof value === 'number' && value < 0) {
    return true;
  }
  return false;
};

/**
  * CategoricalList: Renders a list of categorical bin items
  */
class CategoricalList extends Component {
  constructor(props) {
    super(props);
    this.categoricalListElement = React.createRef();
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    this.colorPresets = [
      getCSSVariableAsString('--cat-color-seq-1'),
      getCSSVariableAsString('--cat-color-seq-2'),
      getCSSVariableAsString('--cat-color-seq-3'),
      getCSSVariableAsString('--cat-color-seq-4'),
      getCSSVariableAsString('--cat-color-seq-5'),
      getCSSVariableAsString('--cat-color-seq-6'),
      getCSSVariableAsString('--cat-color-seq-7'),
      getCSSVariableAsString('--cat-color-seq-8'),
      getCSSVariableAsString('--cat-color-seq-9'),
      getCSSVariableAsString('--cat-color-seq-10'),
    ];
  }

  componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onResize = throttle(() => {
    this.forceUpdate();
  }, 1000 / 60);

  get binSizes() {
    const { bins } = this.props;

    if (!this.categoricalListElement.current) {
      return [];
    }

    const categoricalListElement = this.categoricalListElement.current;

    const bounds = getAbsoluteBoundingRect(categoricalListElement);

    const expandedSize = getExpandedSize(bounds);

    const itemSize = getItemSize(
      bounds,
      bins.length,
      this.isExpanded,
    );

    return {
      expandedSize,
      itemSize,
    };
  }

  get isExpanded() {
    const { expandedBinIndex } = this.props;
    return expandedBinIndex !== null;
  }

  getCatColor = (itemNumber, bin) => {
    if (itemNumber < 0) { return null; }
    const categoryColor = this.colorPresets[itemNumber % this.colorPresets.length];

    if (isSpecialValue(bin.value)) {
      return color(categoryColor).desaturate(0.6).darken(0.5).toString();
    }

    return categoryColor;
  };

  getBinSize = (index) => {
    const { expandedBinIndex } = this.props;
    return expandedBinIndex === index
      ? this.binSizes.expandedSize
      : this.binSizes.itemSize;
  };

  renderCategoricalBins = () => {
    const {
      bins,
      stage,
      prompt,
      activePromptVariable,
      promptOtherVariable,
      getNodeLabel,
      nodeColor,
      onExpandBin,
      expandedBinIndex,
    } = this.props;

    return bins
      .map((bin, index) => (
        <CategoricalListItem
          id={`CATBIN_ITEM_${stage.id}_${prompt.id}_${index}`}
          key={index}
          index={index}
          getNodeLabel={getNodeLabel}
          nodeColor={nodeColor}
          bin={bin}
          size={this.getBinSize(index)}
          activePromptVariable={activePromptVariable}
          promptOtherVariable={promptOtherVariable}
          accentColor={this.getCatColor(index, bin)}
          onExpandBin={onExpandBin}
          isExpanded={expandedBinIndex === index}
          sortOrder={prompt.binSortOrder}
          stage={stage}
        />
      ));
  };

  render() {
    const {
      bins,
      expandedBinIndex,
      onExpandBin,
    } = this.props;

    const listClasses = cx(
      'categorical-list',
      `categorical-list--items--${bins.length}`,
      { 'categorical-list--expanded': expandedBinIndex !== null },
    );

    // Render before filter, because we need to preserve order for colors.
    const categoricalBins = this.renderCategoricalBins();

    const expandedBin = categoricalBins[expandedBinIndex];

    const otherBins = categoricalBins.filter(
      (_, index) => index !== expandedBinIndex,
    );

    return (
      <div
        className={listClasses}
        ref={this.categoricalListElement}
        onClick={(e) => { e.stopPropagation(); onExpandBin(); }}
      >
        <Flipper
          flipKey={expandedBinIndex}
          className="categorical-list__items"
        >
          {expandedBin}
          {otherBins}
        </Flipper>
      </div>
    );
  }
}

CategoricalList.propTypes = {
  activePromptVariable: PropTypes.string.isRequired,
  promptOtherVariable: PropTypes.string,
  bins: PropTypes.array.isRequired,
  getNodeLabel: PropTypes.func.isRequired,
  nodeColor: PropTypes.string.isRequired,
  prompt: PropTypes.object.isRequired,
  stage: PropTypes.object.isRequired,
  expandedBinIndex: PropTypes.number,
  onExpandBin: PropTypes.func.isRequired,
};

CategoricalList.defaultProps = {
  expandedBinIndex: null,
  promptOtherVariable: null,
};

const matchVariable = (node, variable, value) => (
  node[entityAttributesProperty][variable]
  && node[entityAttributesProperty][variable].includes(value)
);

const hasOtherVariable = (node, otherVariable) => (
  otherVariable && node[entityAttributesProperty][otherVariable] !== null
);

const matchBin = (bin, variable) => (node) => (
  matchVariable(node, variable, bin.value) || hasOtherVariable(node, bin.otherVariable)
);

const appendNodesForBin = (nodes, activePromptVariable) => (bin) => ({
  ...bin,
  nodes: nodes.filter(matchBin(bin, activePromptVariable)),
});

function makeMapStateToProps() {
  const getCategoricalValues = makeGetVariableOptions(true);
  const getPromptVariable = makeGetPromptVariable();
  const getStageNodes = makeNetworkNodesForType();

  return function mapStateToProps(state, props) {
    const stageNodes = getStageNodes(state, props);
    const activePromptVariable = getPromptVariable(state, props);
    const [promptOtherVariable, promptOtherVariablePrompt] = getPromptOtherVariable(state, props);
    const getNodeLabel = makeGetNodeLabel();
    const getNodeColor = makeGetNodeColor();
    const bins = getCategoricalValues(state, props)
      .map(appendNodesForBin(stageNodes, activePromptVariable));

    return {
      activePromptVariable,
      promptOtherVariable,
      promptOtherVariablePrompt,
      bins,
      getNodeLabel: getNodeLabel(state, props),
      nodeColor: getNodeColor(state, props),
    };
  };
}

export { CategoricalList };

export default compose(
  connect(makeMapStateToProps),
  MonitorDragSource(['isDragging', 'meta']),
)(CategoricalList);
