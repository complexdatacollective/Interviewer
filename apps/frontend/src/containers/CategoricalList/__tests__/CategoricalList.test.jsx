/* eslint-env jest */

import React from 'react';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import { CategoricalList } from '../CategoricalList';
import { CategoricalListItem } from '../CategoricalListItem';

const mockStore = {
  getState: () => ({
  }),
  subscribe: () => ({
  }),
};

const mockProps = {
  bins: [{ label: 'one', value: 1, nodes: [1, 2, 3] }, { label: 'two', value: 2, nodes: [4, 5] }],
  stage: { id: 1 },
  prompt: { id: 1 },
  getNodeLabel: () => 'label',
  nodeColor: '',
  expandedBinValue: '',
};

describe('CategoricalList component', () => {
  it('renders categorical list', () => {
    const component = shallow(
      <CategoricalList {...mockProps} />,
    );

    expect(component).toMatchSnapshot();
  });

  it('items match the number of bins', () => {
    const component = mount((
      <Provider store={mockStore}>
        <CategoricalList {...mockProps} />
      </Provider>
    ));

    expect(component.find(CategoricalListItem)).toHaveLength(2);
  });
});
