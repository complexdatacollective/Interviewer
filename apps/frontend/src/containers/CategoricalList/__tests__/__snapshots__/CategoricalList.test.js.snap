// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoricalList component renders categorical list 1`] = `
<div
  className="categorical-list categorical-list--items--2"
  onClick={[Function]}
>
  <Flipper
    applyTransformOrigin={true}
    className="categorical-list__items"
    element="div"
    flipKey={null}
  >
    <Connect(CategoricalListItem)
      accentColor={null}
      bin={
        Object {
          "label": "one",
          "nodes": Array [
            1,
            2,
            3,
          ],
          "value": 1,
        }
      }
      getNodeLabel={[Function]}
      id="CATBIN_ITEM_1_1_0"
      index={0}
      isExpanded={false}
      key="0"
      nodeColor=""
      promptOtherVariable={null}
      stage={
        Object {
          "id": 1,
        }
      }
    />
    <Connect(CategoricalListItem)
      accentColor={null}
      bin={
        Object {
          "label": "two",
          "nodes": Array [
            4,
            5,
          ],
          "value": 2,
        }
      }
      getNodeLabel={[Function]}
      id="CATBIN_ITEM_1_1_1"
      index={1}
      isExpanded={false}
      key="1"
      nodeColor=""
      promptOtherVariable={null}
      stage={
        Object {
          "id": 1,
        }
      }
    />
  </Flipper>
</div>
`;
